tket/src/WeightSubgrMono/include/WeightSubgrMono/WeightPruning/WeightNogoodDetectorManager.hpp

```diff
@@ -0,0 +1,172 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <map>
+#include <set>
+
+#include "../GraphTheoretic/GeneralStructs.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+struct FixedData;
+class SearchBranch;
+
+/** Should be attached to a specific search branch.
+ * Uses simple heuristics and feedback control to decide
+ * when to activate the weight nogood detector,
+ * and when to skip (trying to guess if the extra calculation
+ * is worth it).
+ *
+ * It doesn't try to TIME or estimate AMOUNT of calculation
+ * by the detector, or use fancy graph theory. It doesn't really
+ * know anything about detection itself.
+ * It just looks at results: a calculation was done, and a nogood
+ * detected or not.
+ * If not, we know by how much the lower bound fell short
+ * of the required amount, so we can have "OK" or "bad" failures.
+ *
+ * Detecting a nogood is a SUCCESS; it prunes, and we ASSUME
+ * that the time saved by pruning is more than the time spent calculating
+ * to detect the nogood originally.
+ *
+ * Not detecting a nogood is a FAILURE. We spent time calculating,
+ * looking for a nogood but not detecting one (whether or not
+ * it really is a nogood; we just don't know at this stage).
+ *
+ * Really, almost ALL calculations (looking for node reduction,
+ * nogood detection, etc. etc.) should be wrapped inside a manager object.
+ * It would be great to have a general AI-type manager with a few simple
+ * adjustable (or self-adjusting!) parameters to control such things
+ * automatically; but that's probably very hard to do well.
+ */
+class WeightNogoodDetectorManager {
+ public:
+  WeightNogoodDetectorManager();
+
+  ~WeightNogoodDetectorManager();
+
+  /** Decide, based upon the current data and previous results,
+   * whether to try detecting or not.
+   * If this returns true, the caller MUST follow the advice and activate
+   * the detector, and immediately report back the result (see the "register"
+   * functions).
+   * @param current_weight The total weight (scalar product)
+   * @param max_weight The total allowed weight of any solution
+   * @param current_sum_of_p_edge_weights The sum of p-weights currently
+   * assigned
+   * @param total_p_edge_weights The total of all p-weights in any complete
+   * solution.
+   * @param current_number_of_assigned_vertices How many p-vertices are
+   * assigned.
+   * @param current_number_of_unassigned_vertices How many p-vertices are
+   * unassigned.
+   * @return True if the weight nogood detector should be used; if so, the
+   * caller MUST report the result immediately using a "register" function.
+   */
+  bool should_activate_detector(
```

Comment:
Simplicity - weight nogood detector and the manager class are actually totally different, with different responsibilities. The nogood detector calculates a rigorous lower bound on the total weight if a partial solution is extended. The manager decides whether or not to do this calculation in the first place - it is expensive and useless if it doesn't detect a nogood. Note that the manager effectively only uses the nogood detector as a "black box" - it is an automated controller of the nogood detector.

Reply:
Why can't we change the scope of weight nogood detector to be self managing?

Reply:
I think it's usually regarded as good practice to keep classes as independent as possible.

However in this case I suppose if the manager class contains a detector object, it would be clear enough. I'll do that.

Reply:
(Or the other way: the manager contained within the detector, so the caller doesn't see it).

Reply:
In fact, ended up being a new THIRD class which contains both objects.